src/coreclr/vm/jitinterface.cpp

```diff
@@ -7239,8 +7239,34 @@ bool getILIntrinsicImplementationForInterlocked(MethodDesc * ftn,
     if (ftn->GetMemberDef() != CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_T)->GetMemberDef())
         return false;
 
-    // Get MethodDesc for non-generic System.Threading.Interlocked.CompareExchange()
-    MethodDesc* cmpxchgObject = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT);
+    // Determine the type of the generic T method parameter
+    _ASSERTE(ftn->HasMethodInstantiation());
+    _ASSERTE(ftn->GetNumGenericMethodArgs() == 1);
+    TypeHandle typeHandle = ftn->GetMethodInstantiation()[0];
+    MethodTable* methodTable = typeHandle.GetMethodTable();
+
+    MethodDesc* cmpxchg;
+
+    // Based on the generic method parameter, determine which overload of CompareExchange
+    // to delegate to, or if we can't handle the type at all.
+    if (!typeHandle.IsValueType())
+    {
+        cmpxchg = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT);
+    }
+    else if (CorTypeInfo::IsPrimitiveType(typeHandle.GetVerifierCorElementType()))
+    {
+        UINT size = typeHandle.GetSize();
+        _ASSERTE(size == 1 || size == 2 || size == 4 || size == 8);
+        cmpxchg =
+            size == 1 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_BYTE) :
+            size == 2 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_USHRT) :
+            size == 4 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_UINT) :
+            CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_ULONG);
+    }
```

Comment:
I don't think this will be needed if we re-apply the JIT intrinsic, only the object part might slightly help the inliner.

Reply:
Yep, we already discussed this, I think Stephen decided to leave it for separate clean up

Reply:
@jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

Reply:
> @jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

JIT can't handle the object case as intrinsic for most cases due to the need for memory barriers.

Reply:
> @jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

My concern was that the generic method with type switches would be too much work for the inliner and it would produce sub-optimal code when used in real-world code without intrinsic handling. I assume that the codegen diffs captured earlier in this PR confirmed that. @stephentoub Is there a short summary of what you found in https://github.com/dotnet/runtime/pull/104558#discussion_r1669457451 ?

I do not have an opinion about where the intrinsic handling happens exactly.

> JIT can't handle the object case as intrinsic for most cases due to the need for memory barriers (it only handles null and FOH objects there).

Handle as intrinsic != produce code without any calls. There is nothing fundamental preventing all intrinsic IL rewriting done by the VM to be done in the JIT instead. (We may need some new methods on JIT/EE interface.)

Reply:
> My concern was that the generic method with type switches would be too much work for the inliner 

We can  workaround this by, basically, special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget. In theory, it wouldn't be needed if our inliner budget was smart enough to predict foldable branches (and their sizes) precisely.

Reply:
> special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget

Would this special casing also need to include how this method gets imported so that it actually consumes 0 out of JIT inliners budget (e.g. no locals create by the inlining, etc.)?

Reply:
> > special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget
> 
> Would this special casing also need to include how this method gets imported so that it actually consumes 0 out of JIT inliners budget (e.g. no locals create by the inlining, etc.)?

No, but I'd expect most of its body to be trimmed early during importer (all those `if (Unsafe.Sizeof`) so those branches won't be even materialized. The problem that Inliner can't precisely predict it. There might be some extra locals compare to what we see now but I am not sure. I can try to handle this as a follow up if this is merged as is.

Reply:
> @stephentoub Is there a short summary of what you found in https://github.com/dotnet/runtime/pull/104558#discussion_r1669457451 ?

A bunch of cases where calls previously inlined were no longer inlined.